src/Network/WebSockets/TSSocketManager.m

```diff
@@ -218,17 +219,21 @@ - (void)processWebSocketRequestMessage:(WebSocketRequestMessage *)message {
     [self keepAliveBackground];
 
     if ([message.path isEqualToString:@"/api/v1/message"] && [message.verb isEqualToString:@"PUT"]) {
-        NSData *decryptedPayload =
-            [Cryptography decryptAppleMessagePayload:message.body withSignalingKey:TSStorageManager.signalingKey];
-
-        if (!decryptedPayload) {
-            DDLogWarn(@"Failed to decrypt incoming payload or bad HMAC");
-            return;
-        }
+        // SessionCipher state is mutated in a couple places, during decryption, in handleReceivedEnvelope.
+        // To ensure consistent state, that decryption mutation must occur on the same queue as encryption.
+        dispatch_async([OWSDispatch sessionCipher], ^{
+            NSData *decryptedPayload =
+                [Cryptography decryptAppleMessagePayload:message.body withSignalingKey:TSStorageManager.signalingKey];
+
+            if (!decryptedPayload) {
+                DDLogWarn(@"Failed to decrypt incoming payload or bad HMAC");
+                return;
+            }
 
-        OWSSignalServiceProtosEnvelope *envelope = [OWSSignalServiceProtosEnvelope parseFromData:decryptedPayload];
+            OWSSignalServiceProtosEnvelope *envelope = [OWSSignalServiceProtosEnvelope parseFromData:decryptedPayload];
 
-        [[TSMessagesManager sharedManager] handleReceivedEnvelope:envelope];
+            [[TSMessagesManager sharedManager] handleReceivedEnvelope:envelope];
```

Comment:
So it looks like `handleReceivedEnvelope` used to be invoked on the main thread but will now be invoked on the cipher queue.  Why not dispatch async to main thread?  That way:

* Only cipher work will be done on cipher queue.
* Any assumptions about concurrency in `handleReceivedEnvelope` won't be violated.
* Don't have to worry about `handleReceivedEnvelope` being expensive and blocking the cipher queue.